### ISSUE TITLE
commitment_weight

### ISSUE THREAD

**Issue by Levia-Mobius** [EdoardoBotta/RQ-VAE-Recommender#31]
Hi Botta, many thanks for your work. 

I have a question regarding the rqvae.py. When I cheked the code, I found the commitment_weight seems not work in the rqvae.py (line 93):
`self.rqvae_loss = QuantizeLoss(self.commitment_weight)`

I have reviewed the whole rqvae.py but didn't find where it works. However, in the same py file (line: 64~73):
`self.layers = nn.ModuleList(modules=[
            Quantize(
                embed_dim=embed_dim,
                n_embed=codebook_size,
                forward_mode=codebook_mode,
                do_kmeans_init=codebook_kmeans_init,
                codebook_normalize=i == 0 and codebook_normalize,
                sim_vq=codebook_sim_vq
            ) for i in range(n_layers)
        ])`

The quantize loss is actually calculated in this module (refer to Quantize in quantize.py) (line 75) as follows:
`self.quantize_loss = QuantizeLoss(commitment_weight)`

I wonder whether the commitment_weight should be used here cause it is a parameter of Quantize, and the `self.rqvae_loss` in rqvae.py (line 93) doesn't work.

Thanks in advance :)

Best,
Levia

**Comment by EdoardoBotta** [comment]
Hi @Levia-Mobius , thank you for pointing this our. I have pushed a fix to correctly propagate the commitment weight: https://github.com/EdoardoBotta/RQ-VAE-Recommender/commit/a9e1d97529e87a04d4f3d5b063657b0c4e53729a. Feel free to pull the latest commit.

**Comment by Levia-Mobius** [comment]
Many thanks for this quick update!! Another question is about the experiments results. When I integrate this rq-vae into my model structure, I found although the number of duplicated code sequence (e.g., both x1 and x2 has the same code from each codebook: c1 from book1 and c3 from book2) in training set is small (around 10/128, 128 is the batch size), it still shows a high level in the validation set (around 204/1000, 1000 is the validation size). I wonder whether the rq-vae itself has this problem or there may be some wrongs in my model...? I'll appreciate it if I could hear some suggestions from you. Thanks in advance :D

**Comment by Levia-Mobius** [comment]
> Hi [@Levia-Mobius](https://github.com/Levia-Mobius) , thank you for pointing this our. I have pushed a fix to correctly propagate the commitment weight: [a9e1d97](https://github.com/EdoardoBotta/RQ-VAE-Recommender/commit/a9e1d97529e87a04d4f3d5b063657b0c4e53729a). Feel free to pull the latest commit.

Many thanks for this quick update!! Another question is about the experiments results. When I integrate this rq-vae into my model structure, I found although the number of duplicated code sequence (e.g., both x1 and x2 has the same code from each codebook: c1 from book1 and c3 from book2) in training set is small (around 10/128, 128 is the batch size), it still shows a high level in the validation set (around 204/1000, 1000 is the validation size). I wonder whether the rq-vae itself has this problem or there may be some wrongs in my model...? I'll appreciate it if I could hear some suggestions from you. Thanks in advance :D

**Comment by EdoardoBotta** [comment]
Actually 10/128 duplicate ids looks quite high for training as well. That's about 7% of items being assigned to the same id. In my experience with the Amazon dataset, this percentage has been below 1% (You can see it in the max_id_duplicates plot: https://wandb.ai/botta-edoardo-carnegie-mellon-university/rq-vae-training/reports/RQVAE-Training-Beauty--VmlldzoxMTM4MjY4MQ). 

This can be very sensitive on the choice of hyperparameters and on the dataset you are using. With no further details, I would probably suggest trying to tune the dimension and number of vectors in the codebook as a first step.

**Comment by Levia-Mobius** [comment]
> Can you clarify what you mean by 10/128 have duplicate ids? How are these distributed? For example, is it 10 items with the same id or they are only pairwise sharing ids? Actually if 10 items on a batch of 128 all share the same id, and that happens consistently across batches, it looks quite high for training as well. What's the proportion of distinct ids you are seeing in the batch? (You can see it in the p_unique_ids plot that less than 1% : https://wandb.ai/botta-edoardo-carnegie-mellon-university/rq-vae-training/reports/RQVAE-Training-Beauty--VmlldzoxMTM4MjY4MQ).
> 
> This can be very sensitive on the choice of hyperparameters and on the dataset you are using. With no further details, I would probably suggest trying to tune the dimension and number of vectors in the codebook as a first step.

Sorry for my late reply. The 10/128 means in a batch (with size 128), the sum of duplication id sequence number is 10. 
For example, if x1 has the same id sequence with x2  and x3, the duplication of it is recorded as 2; in the same batch, if x 10 is same with x100, then the duplication is 1; so the whole duplication number of this batch is 3 = 2+1

I'm still working on this part now :( I set each codebook size as 80 cause there are only around 21000 data points in the training set. Do u think this codebook size is too small? But 3 codebooks with size 80 should represent around 512,000 (80^3) data points theoretically... I have no idea why the duplication rate in validation set still has 6%~7%...

Sincerely look forward to your suggestions :)

**Comment by Levia-Mobius** [comment]
> Can you clarify what you mean by 10/128 have duplicate ids? How are these distributed? For example, is it 10 items with the same id or they are only pairwise sharing ids? Actually if 10 items on a batch of 128 all share the same id, and that happens consistently across batches, it looks quite high for training as well. What's the proportion of distinct ids you are seeing in the batch? (You can see it in the p_unique_ids plot that less than 1% : https://wandb.ai/botta-edoardo-carnegie-mellon-university/rq-vae-training/reports/RQVAE-Training-Beauty--VmlldzoxMTM4MjY4MQ).
> 
> This can be very sensitive on the choice of hyperparameters and on the dataset you are using. With no further details, I would probably suggest trying to tune the dimension and number of vectors in the codebook as a first step.

I just reviewed the data processing py script, it seems that only movie title is used to train the RQ-VAE for MovieLens dataset, and p_unique_ids is calculated on the whole dataset with model.eval(). Is my understanding correct? Thanks!

**Comment by EdoardoBotta** [comment]
Yeah, your understanding is correct. I would advise sticking to the Amazon datasets, unfortunately movielens does not have enough text to produce text embeddings for items.

**Comment by qiancijun** [comment]
> > Can you clarify what you mean by 10/128 have duplicate ids? How are these distributed? For example, is it 10 items with the same id or they are only pairwise sharing ids? Actually if 10 items on a batch of 128 all share the same id, and that happens consistently across batches, it looks quite high for training as well. What's the proportion of distinct ids you are seeing in the batch? (You can see it in the p_unique_ids plot that less than 1% : https://wandb.ai/botta-edoardo-carnegie-mellon-university/rq-vae-training/reports/RQVAE-Training-Beauty--VmlldzoxMTM4MjY4MQ).
> > This can be very sensitive on the choice of hyperparameters and on the dataset you are using. With no further details, I would probably suggest trying to tune the dimension and number of vectors in the codebook as a first step.
> 
> I just reviewed the data processing py script, it seems that only movie title is used to train the RQ-VAE for MovieLens dataset, and p_unique_ids is calculated on the whole dataset with model.eval(). Is my understanding correct? Thanks!

Have you ever encountered the problem of unbalanced distribution of the codebook? The codebook shows a long-tail distribution, and the content output by the decoder tends to be consistent. Thanks ^^
